app/grandchallenge/algorithms/models.py

```diff
@@ -619,12 +647,148 @@ def spent_credits(self, user):
         )
 
 
+def algorithm_models_path(instance, filename):
+    return (
+        f"models/"
+        f"{instance._meta.app_label.lower()}/"
+        f"{instance._meta.model_name.lower()}/"
+        f"{instance.pk}/"
+        f"{get_valid_filename(filename)}"
+    )
+
+
+class AlgorithmModel(UUIDModel):
+    creator = models.ForeignKey(
+        settings.AUTH_USER_MODEL, null=True, on_delete=models.SET_NULL
+    )
+    algorithm = models.ForeignKey(
+        Algorithm, on_delete=models.PROTECT, related_name="algorithm_models"
+    )
+    import_status = models.PositiveSmallIntegerField(
+        choices=ImportStatusChoices.choices,
+        default=ImportStatusChoices.INITIALIZED,
+        db_index=True,
+    )
+    status = models.TextField(editable=False)
+    user_upload = models.ForeignKey(
+        UserUpload,
+        blank=True,
+        null=True,
+        on_delete=models.SET_NULL,
+        validators=[validate_gzip_mimetype],
+    )
+    model = models.FileField(
+        blank=True,
+        upload_to=algorithm_models_path,
+        validators=[ExtensionValidator(allowed_extensions=(".tar.gz",))],
+        help_text=(
+            ".tar.gz file of the algorithm model that will be extracted to /opt/ml/model/ during inference"
+        ),
+        storage=private_s3_storage,
```

Comment:
This should be on the protected storage as we will need to serve them later, see https://github.com/DIAGNijmegen/rse-roadmap/issues/301.